tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberSetMethod.scala

```diff
@@ -0,0 +1,121 @@
+
+  override val methodName: Invocation.MethodName = MethodName("TeamMailboxMember/set")
+
+  override val requiredCapabilities: Set[CapabilityIdentifier] = Set(JMAP_CORE, JMAP_MAIL, LINAGORA_TEAM_MAILBOXES)
+
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, TeamMailboxMemberSetRequest] =
+    Serializer.deserializeSetRequest(invocation.arguments.value)
+      .asEither.left.map(ResponseSerializer.asException)
+
+  override def doProcess(capabilities: Set[CapabilityIdentifier],
+                         invocation: InvocationWithContext,
+                         mailboxSession: MailboxSession,
+                         request: TeamMailboxMemberSetRequest): Publisher[InvocationWithContext] =
+    update(mailboxSession.getUser, request)
+      .map(response => Invocation(
+        methodName = methodName,
+        arguments = Arguments(Serializer.serializeSetResponse(response).as[JsObject]),
+        methodCallId = invocation.invocation.methodCallId))
+      .map(InvocationWithContext(_, invocation.processingContext))
+
+  private def update(username: Username, request: TeamMailboxMemberSetRequest): SMono[TeamMailboxMemberSetResponse] =
+    SFlux.fromIterable(request.validatedUpdateRequest())
+      .flatMap(parsingRequestResult => update(username, parsingRequestResult))
+      .collectSeq()
+      .map(listResult => TeamMailboxMemberSetResponse.from(request.accountId, listResult))
+
+  private def update(username: Username,
+                     parsingRequestResult: ParsingRequestResult): SMono[TeamMailboxMemberSetResult] =
+    parsingRequestResult match {
+      case ParsingRequestFailure(teamMailboxNameDTO, exception) => SMono.just(TeamMailboxMemberSetResult.notUpdated(teamMailboxNameDTO, SetErrorDescription(exception.message)))
+      case ParsingRequestSuccess(teamMailbox, membersUpdateToAdd, membersUpdateToRemove) =>
+        checkIfAnyUsersDoesNotExist(membersUpdateToAdd, membersUpdateToRemove)
+          .collectSeq()
+          .filter(usernames => usernames.nonEmpty)
+          .map(usernames => TeamMailboxMemberSetResult.notUpdated(teamMailbox,
+            SetErrorDescription(s"Some users do not exist in the system: ${usernames.map(username => username.asString()).toArray.mkString("", ", ", "")}")))
+          .switchIfEmpty(updateTeamMailboxMembers(username, teamMailbox, membersUpdateToAdd, membersUpdateToRemove))
+          .onErrorResume {
+            case _: TeamMailboxNotFoundException => SMono.just(TeamMailboxMemberSetResult.notUpdated(teamMailbox, SetErrorDescription("Team mailbox is not found or not a member of the mailbox")))
+            case _: Throwable => SMono.just(TeamMailboxMemberSetResult.notUpdated(teamMailbox, SetErrorDescription("Internal error")))
+          }
+    }
+
+  private def checkIfAnyUsersDoesNotExist(membersUpdateToAdd: List[TeamMailboxMember],
```

Comment:
I spent 5 minutes just understanding this does listNonExisting users.

Please use good names.

Reply:
and return `SMono[Seq[Username]]` 
Then, at line 61 it will be pretty

```scala
listNonExistingUsers(membersUpdateToAdd, membersUpdateToRemove)
  .flatMap {
    case Seq.empty => updateTeamMailboxMembers(username, teamMailbox, membersUpdateToAdd, membersUpdateToRemove)
    case unExistUsernames => SMono.just(TeamMailboxMemberSetResult.notUpdated(teamMailbox,
      SetErrorDescription(s"Some users do not exist in the system: ${unExistUsernames.map(username => username.asString()).toArray.mkString("", ", ", "")}")))
  }
}
```